src/psyclone/psyir/symbols/symbol_table.py

```diff
@@ -303,6 +302,40 @@ def deep_copy(self):
                                      routine.from_container))
             symbol.routines = new_routines
 
+        # Ensure any Symbols referenced in array shapes are also updated
```

Comment:
In fact, not for this PR but this method may also be useful for nodes to reattach them to scopes with different symbols by using the current names. Right now we just reattach thing hoping that symbols are there without any guarantees.

Reply:
Yes, it might also be useful/re-usable in inlining. I've done your `relink` suggestion and it works out quite nicely (although I had to add it to `Node` too). Currently the `Node.relink` uses `walk` to find References but obviously this will miss (at least) precision symbols in Literals. I remember we've got at least one TODO to switch to using reference_accesses instead to avoid this problem. Do you think I should do that here? Also, I'm thinking that perhaps `replace_symbols` or `update_symbols` might be a more self-explanatory name rather than `relink`. What do you think?

Reply:
The issue for reference_accesses is #2542 and that depends on #2424 (make it work for nested scopes). That's not going to happen any time soon I think (I previously looked at doing #2424 and it got very difficult). There are a few possible approaches to replacing Symbols in a PSyIR tree:
 * Add a `relink` method to Node and have it look for all References (and Literals) and replace their symbols;
 * Add a `relink` method to Node that simply walks the tree calling `relink`. Then implement the necessary functionality in `relink` methods for `Reference` and `Literal` classes (at least).
 * Create a Visitor with handlers for `Reference`, `Literal`, `Routine`...

Reply:
I do not mind the name, in fact linking is probably too broad for Node, as there are several "links". Also for Node it does not need a symbol_table parameter, as to avoid the same problems with nested scopes it should you a lookup from the scope it is connected in. Maybe "update_symbols_to_current_scope" or "bind_symbols_to_current_scope"? In symbols it can be "bind_symbols_to(symbol_table)", or something similar?

Regarding the implementation, I see you started with the Visitor, but I am unsure about this choice. A Visitor still decouples the implementation for node, which is good for having a multiple backends but when adding mandatory functionality that depends on the attributes of a class, I think its clear if its all in the same place, with a specialised method in this class (your second bullet point). This way, when some body adds or removes an attribute from a node, it has to check if it needs to update the eq, copy, relink or that class (instead of knowing how other files may be affected). Also visitors do not do that well with inheritance. While a overwiting method only has to care about its attributes, and then calling the super() version, which will care about theirs. But I am happy to wait for the review to have a better opinion if you prefer the visitor.